analyzer-java/src/main/java/com/infosupport/ldoc/analyzerj/AnalysisVisitor.java

```diff
@@ -78,13 +79,27 @@ public AnalysisVisitor(SymbolResolver resolver) {
     this.resolver = resolver;
   }
 
-  /** Resolves the given type to a String with its fully-qualified class name. */
-  private String resolve(Type type) {
+  /** Resolves the given node to a String with its fully-qualified class name.
+   * Calls the corresponding resolver method based on the type of the node.
+   * */
+  private String resolve(Node node) {
```

Comment:
I'm alright with changing `this.resolve(something)` to be a kind of "look at what something is, and do something sensible based on what it is exactly" method, but I would prefer to do the switching on type by way of overloading:

```
private String resolve(Type t) {
  ...
}

private String resolve(AnnotationExpr ae) {
  ...
}

private String resolve(Expression e) {
  ...
}
```

This is a Visitor, after all - that design pattern is intended to remove the need for `instanceof` checks.

This would also remove the need for the LdocException.

Reply:
I considered your suggestion as well. The main disadvantage of this would be code duplication.

As we work around bug(s) in JavaParser, there is a bit logic that we catch some runtime exceptions and in that case, return a placeholder string. As this logic depends on issues in JavaParser, we rightfully added some comments referring to this issue. In case of method overloading, there would be three methods with duplicates of the same try/catch mechanism, argumentation/reference comment, and 'unknown placeholder' return statements.

We also have to consider that the original resolve method was defined as a workaround for the first encountered bug in JavaParser. At that time, some places in the visitor referred to the resolve method, and some still called the JavaParser directly. In that sense, the resolve method is more of a wrapper to the JavaParser resolver than it is part of the visitor. Following this, if one desires component separation, we could explicitly define a wrapper for the JavaParser resolver that works around the bugs in such a way that it works for LivingDocumentation.


All in all, I still think there are reasons to keep it this way or change it to an entirely different structure altogether,
I am curious whether my argumentation was able to persuade you or whether you still think differently :)

Reply:
I am not worried about code duplication in this case, especially since this is, as you say, code that can be removed once JavaParser no longer has this bug.

Reply:
I think this mechanism will be in place longer than we might hope.

Currently, the record bug in JavaParser has already been open for a long time and manifests in different ways, with no indication that it will be fixed soon. 

I think it comes down to the issue of how we want to handle JavaParser not being able to resolve certain types. Do we want to fail the analyzer, or do we at all times want to provide a result (with "?" in place of the type)?
In the case of the latter, it might be good to just permanently keep the try/catch clause (maybe even catching generic runtime exceptions and log a warning)

Reply:
Sure. We would keep the try/catch clauses as long as JavaParser has this limitation. If JavaParser has the limitation until the end of time, we would keep the try/catch clause until the end of time.

In the meantime, my preference would be to fall back on the unqualified name if possible, I think that is the most likely to be useful. It is also the existing behavior.

Reply:
> I tried to say it might be good to keep the try/catch clause even if all the bugs we currently work around are fixed.

I agree! But that doesn't really affect this PR, right? We can make this decision when that time comes (which, as you said, may not happen soon or may not happen at all). At that point we would reconsider: do we still want these `catch`es? What use is it to us? And if we do want to keep it, we could even think about: are we happy enough with it in its current shape, so with the three overloads, or do we instead want the shape in this PR with the `instanceof` checks and the `throw`, or do we want something else entirely?

But that's then and this is now. For now, falling back on a fixed value seems fine to me, I just don't like the `if` `instanceof` tree.

> For the sake of consistency I changed it to an explicit 'unknown type' ["?" everywhere].

I'm okay with this; consistency is also important.

Reply:
> But that's then and this is now. For now, falling back on a fixed value seems fine to me, I just don't like the `if` `instanceof` tree.

I agree with the fact that the if statement tree is not that pleasing. However, I still think it would be nice to keep this workaround logic in one place as opposed to three separate methods. Especially in case we plan to change the logic, but also in case it stays this way.

For now, if desired, the resolve method can be split into three methods. For the moment, I at least removed the custom LdocException.

Reply:
Yes, please do the split; it's the last thing that's keeping me from approving this PR, otherwise this looks good.